### PR TITLE
Add javascript-extension

### PR DESCRIPTION
**Pull request by gnestor**
Addresses https://github.com/jupyterlab/jupyterlab/issues/3118 and https://github.com/jupyterlab/jupyterlab/issues/3748#issuecomment-390529500

javascript-extension

Demo notebook: https://gist.github.com/gnestor/f1893e0226ced227e910f11b769adc06

### CODE REVIEW

File: packages/javascript-extension/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+}
+
+/**
+ * A mime renderer factory for text/javascript data.
+ */
+export
+const rendererFactory: IRenderMime.IRendererFactory = {
+  safe: true,
```
Comment (vidartf) [inline packages/javascript-extension/src/index.ts]: I think this should be `false`, but I've misunderstood this before (@sccolbert).

Comment (vidartf) [reply packages/javascript-extension/src/index.ts]: Ref: https://github.com/jupyterlab/jupyterlab/blob/b36f9709112576700e3bd56a6a265a63201e3510/packages/rendermime/src/factories.ts#L85-L94

**Comment by ian-r-rose** [comment]
Thanks for taking this on @gnestor. A couple of thoughts:
1. We should probably remove the commented-out code in `@jupyterlab/rendermine` in favor of this.
2. We should make a decision about whether to disable this extension by default or not.

**Comment by gnestor** [comment]
@ian-r-rose Ok, good suggestions. 

Regarding the commented code in rendermime, what should we do about `text/html` output? Do we want to completely disallow it? Add a setting to the settings editor (e.g. enabled/disabled)? 

I think whatever we decide for HTML output, we should probably do the same for javascript output. I vote for client-side opt-in vs. labconfig to make it more accessible to users. 

Another option is a cell-based opt-in: 



This very quickly becomes cumbersome (to click run for each cell, especially if you're in a "Run all" situation), but it's also very straight-forward for the user. We could also have a notebook-based opt-in?

**Comment by ellisonbg** [comment]
Looks like a good approach. In a call we talked about adding global configuration for it.

File: packages/javascript-extension/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+   // tslint:disable-next-line
+  return eval(code);
+}
+
+export class ExperimentalRenderedJavascript extends RenderedJavaScript {
+  render(model: IRenderMime.IMimeModel): Promise<void> {
+    const renderJavascript = () => {
+      try {
+        const data = model.data[this.mimeType] as string;
+        evalInContext(data, this.node, document, window);
```
Comment (vidartf) [inline packages/javascript-extension/src/index.ts]: Here you eval the code before trust is checked. I would move the block checking trust to the top of this function.

Comment (vidartf) [reply packages/javascript-extension/src/index.ts]: Oh, yes, I see that now..

**Comment by gnestor** [comment]
Now we have support for HTML output:

**Comment by gnestor** [comment]
@timkpaine To clarify, the "Run" prompt is only displayed if the output is not trusted, so if you authored the notebook or have run the cell once before, then you will not be presented with the prompt. If this proves to be annoying for users, then we will provide a "Remember my settings for this notebook/user" checkbox that will persist in the JupyterLab settings.

**Comment by timkpaine** [comment]
@gnestor so each time the notebook is open I need to trust it? My custom contents manager automatically trusts the output, since we're in a controlled env, but this may be annoying for normal users.

**Comment by gnestor** [comment]
@timkpaine Nope. If you author a notebook, then it's already trusted and you're never prompted. If you download someone else's notebook, you will prompted once for each cell that contains javascript, after which point it's trusted and you will never be prompted again.

**Comment by gnestor** [comment]
@ellisonbg This is ready 👍

**Comment by mpilosov** [comment]
just to confirm, once this is merged (if this is merged...), will sliders with callbacks finally be able to update plots they are supposed to generate? Seems like this is the PR I've been looking for...

**Comment by timkpaine** [comment]
You could already use sliders if you used the ipywidgets ones

**Comment by mpilosov** [comment]
The sliders show up and successfully reflect changes to their values, but the callback functions attached to them that update a plot (using “clear_output” and “display” from ipython) have no effect.

This problem shows up in a few issues/threads, and was the first (so far only) major thing that didn’t naturally transition from jupyter notebooks to lab.


Not sent from my computer.

On May 23, 2018, at 08:46, Tim Paine <notifications@github.com<mailto:notifications@github.com>> wrote:


You could already use sliders if you used the ipywidgets ones

**Comment by jasongrout** [comment]
> just to confirm, once this is merged (if this is merged...), will sliders that trigger callbacks finally be able to update the (matplotlib) plots they are supposed to help generate/modify? Seems like this is the PR I've been looking for...

An ipywidgets slider not updating a `%inline` matplotlib plot displayed in an ipywidget output widget will (very likely) not be solved by this issue. Likely the problem in that case lies somewhere else.

**Comment by zahachtah** [comment]
Does this only work with python kernel? How could I invoke javascript access e.g. running julia on jupyter lab?

**Comment by blink1073** [comment]
Hi @zahachtah, I'm not very familiar with Julia, but it looks like this is what you want: https://github.com/JuliaLang/IJulia.jl/blob/283bba19cc68b516f8cb748485ab234fe5bfc32b/src/inline.jl#L67

**Comment by shashi** [comment]
@zahachtah 

```julia
HTML("<script>alert('hi')</script>")
```

Seems to just work in the latest version of JupyterLab (with Julia).

**Comment by gnestor** [comment]
To answer your question @zahachtah, the mimerender extensions (e.g. JSON, GeoJSON, Vega), including the Javascript extension, work with all kernels. To render as JSON, for example, the kernel must send a [display_data message](https://ipython.readthedocs.io/en/3.x/notebook/nbformat.html#display-data) with a `application/json` output in the output mime bundle. Each of these extensions is associated with a different mime type (e.g. `application/geo+json`, `application/vnd.vega.v3+json`).

**Pull request merged**